### PR TITLE
billing - Add card

### LINKED ISSUE #1368: User should be able to add a card to their account
A user should be able to add a card to their account.

The card details should be sent to Stripe API directly, and the token received from Stripe should be submitted to the API

### PR DESCRIPTION
**Pull request by tanmoyAtb**
closes #1368 

make sure to add to `.env`
`REACT_APP_STRIPE_PK=pk_test_1CtYS7zHbojtudFYMgCT2xcF00eijx1KCE`

Notes: 
Removed some old design code for purchase plan.
Moved some files around under settings!

### CODE REVIEW

**Comment by RyRy79261** [comment]
Looks good, only things to mention my side is: 

It would be nice to have the auto `-` inserting formatting added like the date section 

One thing that, when I see it on paygates, just makes me feel better, is after the first 4 digits it indicates if the cards is a master card or a visa or something, might be out of scope but thought to mention

**Comment by RyRy79261** [comment]
image

**Review by RyRy79261 — APPROVED** [review]

**Comment by tanmoyAtb** [comment]
> Looks good, only things to mention my side is:
> 
> It would be nice to have the auto `-` inserting formatting added like the date section
> 
> One thing that, when I see it on paygates, just makes me feel better, is after the first 4 digits it indicates if the cards is a master card or a visa or something, might be out of scope but thought to mention

Hey @RyRy79261 , so the formatting and card type indication both are there!
try `4242 4242 4242 4242` or `5555 5555 5555 4444` ,
Its just starting with `1234` didn't refer to any card, so it didn't format it either!

**Comment by tanmoyAtb** [comment]
> > Looks good, only things to mention my side is:
> > It would be nice to have the auto `-` inserting formatting added like the date section
> > One thing that, when I see it on paygates, just makes me feel better, is after the first 4 digits it indicates if the cards is a master card or a visa or something, might be out of scope but thought to mention
> 
> Hey @RyRy79261 , so the formatting and card type indication both are there!
> try `4242 4242 4242 4242` or `5555 5555 5555 4444` ,
> Its just starting with `1234` didn't refer to any card, so it didn't format it either!

On another thought, I'll add a default formatting if there's no card type match!

**Review by FSM1 — APPROVED** [review]

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionPlan/AddCardModal.tsx [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+  Button,
+  Grid,
+  TextInput,
+  Typography,
+  useToaster
+} from "@chainsafe/common-components"
+import {
+  createStyles,
+  makeStyles
+} from "@chainsafe/common-theme"
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionPlan/AddCardModal.tsx]:
```suggestion
import { Button, Grid, TextInput, Typography, useToaster } from "@chainsafe/common-components"
import { createStyles, makeStyles } from "@chainsafe/common-theme"
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionPlan/AddCardModal.tsx [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+interface ICreateFolderModalProps {
+  isModalOpen: boolean
+  onClose: () => void
+}
+
+const CreateFolderModal: React.FC<ICreateFolderModalProps> = ({
+  isModalOpen,
+  onClose
+}: ICreateFolderModalProps) => {
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionPlan/AddCardModal.tsx]:
```suggestion
const CreateFolderModal = ({i sModalOpen, onClose }: ICreateFolderModalProps) => {
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionPlan/AddCardModal.tsx [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+    cardExpiry: "",
+    cardCvc: ""
+  })
+  const [cardName, setCardName] = useState("")
+  const [error, setError] = useState("")
+  const [loading, setLoading] = useState(false)
+  const { filesApiClient } = useFilesApi()
+  const { addToastMessage } = useToaster()
+
+  const onCloseModal = () => {
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionPlan/AddCardModal.tsx]: please use `useCallback`

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionPlan/AddCardModal.tsx [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    setCardInputs({ cardNumber: "", cardExpiry: "", cardCvc: "" })
+    setCardName("")
+    setError("")
+    onClose()
+  }
+
+  const onSubmitCard = (e: React.FormEvent<HTMLFormElement>) => {
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionPlan/AddCardModal.tsx]: please use `useCallback`

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionPlan/AddCardModal.tsx [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+            item
+            xs={12}
+            sm={12}
+          >
+            <TextInput
+              value={cardName}
+              onChange={(val) =>
+                setCardName(val?.toString() || "")
+              }
+              // className={classes.cardNumber}
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionPlan/AddCardModal.tsx]:
```suggestion
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionPlan/index.tsx [hunk 6]
```diff
@@ -1,10 +1,35 @@
...
+    container: {
+      margin: constants.generalUnit * 4
+    }
+  })
+)
 
 const PlanView: React.FC = () => {
+  const classes = useStyles()
+
+  const [isAddCardModalOpen, setIsAddCardModalOpen ] = useState(false)
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionPlan/index.tsx]:
```suggestion
  const [isAddCardModalOpen, setIsAddCardModalOpen ] = useState(false)
  
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionPlan/index.tsx [hunk 7]
```diff
@@ -1,10 +1,35 @@
...
   return (
-    <CurrentProduct />
+    <div className={classes.root}>
+      <CurrentProduct />
+      <div className={classes.container}>
+        <Button onClick={() => setIsAddCardModalOpen(true)}>Add Card</Button>
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionPlan/index.tsx]: Translation is missing
```suggestion
        <Button onClick={() => setIsAddCardModalOpen(true)}>
          <Trans>Add Card</Trans>
        </Button>
```

File: packages/files-ui/src/Components/Modules/Settings/SubscriptionPlan/AddCardModal.tsx [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+    }).then((resp: AxiosResponse<{ card: {id: string}}>) => {
+      // send stripe token to API
+      filesApiClient.addCard({ token: resp.data.card.id })
+        .then(() => {
+          onCloseModal()
+          addToastMessage({
+            message: t`Card added successfully`
+          })
+        }).catch(() => {
+          setError(t`Something went wrong`)
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/Settings/SubscriptionPlan/AddCardModal.tsx]: We should be a little more helpful to our users and ask to try again maybe?
Also to help us, logging in the console could help
```suggestion
        }).catch((e) => {
          setError(t`Something went wrong, please try again`)
          console.error(e)
```

**Review by Tbaut — COMMENTED** [review]
Certainly from another PR (that got merged) but we should redirect to the settings/plan page when clicking on "Back to plan settings"

Looks pretty good otherwise

**Comment by tanmoyAtb** [comment]
> Certainly from another PR (that got merged) but we should redirect to the settings/plan page when clicking on "Back to plan settings"
> 
> Looks pretty good otherwise

Updated the path on `back to plan settings` ! and resolved the feedbacks ! thanks @Tbaut

**Comment by Tbaut** [comment]
Thanks @tanmoyAtb I'm afk right now but I think the change in the route setting (removing the `:path`) will break the usage of the built in `useParams` from react router that is used in the tab management. I'll point to the exact line when I'm back if you haven't got it.

Edit : ah you actually didn't remove it, so nothings broken! There's a place in the tab management where the settings/path is set manually, I'll check it out when I'm back in front of the computer

**Comment by Tbaut** [comment]
@tanmoyAtb I had to add some generics to make it type safe but I ended up being able to do what I wanted. Please take a look and merge if you think it's ok.

**Comment by tanmoyAtb** [comment]
> @tanmoyAtb I had to add some generics to make it type safe but I ended up being able to do what I wanted. Please take a look and merge if you think it's ok.

Thanks the types look good to me !

**Pull request merged**